### ISSUE TITLE
goimports only updates standard library imports

### ISSUE THREAD

**Issue by nathj07** [Microsoft/vscode-go#1427]
OS: MacOS Sierra
VS-Code: 1.19.1
Extension: 0.6.70

As discussed under issue #1419 version 0.6.70 only applies `goimports` to standard library packages.

For example:
```go

package main

import(
    "fmt"
)

type data struct {
    Counter int `json:"counter"`
    Desc string `json:"desc"`
}
var input = []byte(`{"counter":42, "desc":"meaning of life"}`)
func main() {
    fmt.Println("Up and running") 
    output := &data{}
    err := json.Unmarshal(input, output) // this will be imported
    if err != nil {
        panic(err)
    }
    spew.Dump(output) // this will not be imported
}
```
This example relies on you having previously run `go get github.com/davecgh/go-spew/spew`  When formatOnSave runs this code will have `encoding/json` added to the list of imports but not `github.com/davecgh/go-spew/spew`.

Thank you for helping out on this

**Comment by ramya-rao-a** [comment]
@nathj07 Just tried your example. Import statement for the `spew` pkg does get added for me. Can you try the below:

- Share any errors in the console when you get this issue. Help -> Toggle Developer Tools -> Console
- Can you try setting `go.formatTool` to `goreturns`. Try to repro the issue. Then change it to `goimports` and  try again to get a repro. See if there is a difference when using the 2 tools or both give you the same problem
- Do you see the issue when manually formatting the file? Right click -> Format Document

**Comment by nathj07** [comment]
I will get to this for you. Right now I'm on family holiday and trying to stay off the computer. I'll get to it early next week. Hope that's ok.

**Comment by ramya-rao-a** [comment]
That's totally fine @nathj07 
Enjoy your holidays

**Comment by nathj07** [comment]
@ramya-rao-a  that is odd; it does in fact work for me even after moving to v 0.6.70of the Go plugin.

However, on more complex code it does not and unfortunately I get nothing in the output pane. If I right-click and `Go: Add Import` it puts the imports in.

I tried with `goreturns` and it turns out I didn't have it installed. Installing that and saving my package with missing imports I get lots of output about `undeclared name` as the required package is not imported. However the import doesn't get added.

As an example for this:
```
package ldapauth

import (
	"context"
	"errors"
	"fmt"
	// "strings"

	// "github.com/Sirupsen/logrus"
	// "github.com/mavricknz/ldap"
)

var (
	// ErrUserNotFound means that the requested user could not be found in the system.
	ErrUserNotFound = errors.New("requested user could not be found")

	// ErrInvalidCredentials means that the password provided for the was not valid. Depending on
	// the nature of the users of the system, it might not be a good idea to indicate that it
	// was merely the password that was wrong, therefore giving the user the information that
	// the username exists.
	ErrInvalidCredentials = errors.New("the login credentials were not valid")
)

// Authenticator is an interface for abstracting out what it means to autheticate a username and
// password.
type Authenticator interface {
	// Authenticate checks to see if the username and password are a valid pair in the given system. If there
	// is a problem authenticating it will return an error. If the user is succesfully authenticated then a
	// list of groups to which that user belongs will be returned.
	Authenticate(ctx context.Context, username, password string) (groups []string, err error)
}

// FakeAuthenticator maps valid usernames to a list of groups to which they belong. It does
// not validate a password.
type FakeAuthenticator map[string][]string

// Authenticate looks to see if username exists in its hash of users to groups and return the list of groups if
// the user is found. If the username is not found, ErrUserNotFound is returned. Password is ignored.
func (fa FakeAuthenticator) Authenticate(ctx context.Context, username, password string) ([]string, error) {
	groups, ok := fa[username]
	if !ok {
		return nil, ErrUserNotFound
	}
	return groups, nil
}

// Config defines the data needed to configure a ldap authenticator.
type Config struct {
	Host         string `yaml:"host"`
	Port         int    `yaml:"port"`
	BaseDN       string `yaml:"base_dn" envconfig:"base_dn"`
	SearchUser   string `yaml:"search_user" envconfig:"search_user"`
	SearchPasswd string `yaml:"search_password" envconfig:"search_password"`
}

// New returns a new Authenticator used to authenticate against an LDAP server as
// described by config.
func New(config Config) *LDAPAuthenticator {
	return &LDAPAuthenticator{
		config: config,
	}
}

// LDAPAuthenticator is used to authenticate against an LDAP server.
type LDAPAuthenticator struct {
	config Config
}

// Authenticate checks to see that username exists (as defined by a search for sAMAccountName=username) and then
// authenticates that user against the provided password. If successful all groups to which the user belongs will
// be returned. Please note that groups only returns the group name, not the entire DN of the group (i.e if the DN
// of the group looks like CN=[GROUP], OU=...., only [GROUP] is returned.
func (la *LDAPAuthenticator) Authenticate(ctx context.Context, username, password string) (groups []string, err error) {
	conn := ldap.NewLDAPConnection(la.config.Host, uint16(la.config.Port))
	err = conn.Connect()
	if err != nil {
		return nil, fmt.Errorf("failed to connect to LDAP server: %v", err)
	}
	defer conn.Close()

	err = conn.Bind(la.config.SearchUser, la.config.SearchPasswd)
	if err != nil {
		return nil, err
	}

	result, err := conn.Search(ldap.NewSimpleSearchRequest(la.config.BaseDN, ldap.ScopeWholeSubtree,
		fmt.Sprintf("(sAMAccountName=%s)", username), []string{"memberOf"}))
	if err != nil {
		return nil, fmt.Errorf("failed to search for user %v, error: %v", username, err)
	}

	if len(result.Entries) < 1 {
		return nil, ErrUserNotFound
	}
	userDN := result.Entries[0].DN

	err = conn.Bind(userDN, password)
	if err != nil {
		return nil, ErrInvalidCredentials
	}

	groups = nil
	for _, group := range result.Entries[0].GetAttributeValues("memberOf") {
		if !strings.HasPrefix(group, "CN=") {
			logrus.WithField("ldap_group", group).Warn("group does not start with 'CN=' as expected")
			continue
		}

		endIndex := strings.Index(group, ",OU=")
		if endIndex < 0 {
			logrus.WithField("ldap_group", group).Warn("group does not contain ',OU=' as expected")
			continue
		}
		groups = append(groups, group[3:endIndex])
	}

	return groups, nil
}
```
For me none of the imports that are commented out there are put in on save using either goreturns or goimports and I see no output.

**Comment by ramya-rao-a** [comment]
I'll try your example in some time. Meanwhile does manual formatting add the missing imports?

**Comment by nathj07** [comment]
yeah, a manual format works if I right-click and Go: Add Import it puts the imports in.

**Comment by ramya-rao-a** [comment]
So you are saying manual format adds the imports? Or you have to run `Go: Add Import` to add the imports?

**Comment by nathj07** [comment]
I see the confusion here. If I use the right-click menus and select `Go: Add Imports` it works and if I run goimports manually on the command line it works. It's just the format on save that is not updating the imports

**Comment by ramya-rao-a** [comment]
What I am asking is independent of `Go: Add Imports`.
Do the missing imports get added when you right-click and choose `Format Document`?

**Comment by nathj07** [comment]
My mistake. 

formatter: goimports
The imports are not added and the output shows only go vet being run

formatter: goreturns
The imports are added

This is using `Format Document`. Format on save does not add imports with either formatter

**Comment by ramya-rao-a** [comment]
Ok, First lets tackle `goimports` not adding the imports when running `Format Document`.
Do you see any errors in the console when running `Format Document`  when using `goimports`?

**Comment by nathj07** [comment]
The output is that `go vet` completes, and that the code doesn't build because of missing imports.

**Comment by ramya-rao-a** [comment]
@nathj07 When I talk about "errors in the console", I mean the console that appears when you go through the Help -> Toggle Developer Tools -> Console and not the output pane where the results of build/vet/lint appear.

**Comment by nathj07** [comment]
With `goimports` as the formatter:
```
[Extension Host] (node:6550) DeprecationWarning: os.tmpDir() is deprecated. Use os.tmpdir() instead.
t.log @ console.ts:123
t._logExtensionHostMessage @ extensionHost.ts:391
(anonymous) @ extensionHost.ts:211
emitTwo @ events.js:106
emit @ events.js:194
process.nextTick @ internal/child_process.js:766
_combinedTickCallback @ internal/process/next_tick.js:73
_tickCallback @ internal/process/next_tick.js:104
console.ts:123 [Extension Host] ApplicationInsights:Sender [Object]
t.log @ console.ts:123
t._logExtensionHostMessage @ extensionHost.ts:391
(anonymous) @ extensionHost.ts:211
emitTwo @ events.js:106
emit @ events.js:194
process.nextTick @ internal/child_process.js:766
_combinedTickCallback @ internal/process/next_tick.js:73
_tickCallback @ internal/process/next_tick.js:104
```

And with `goreturns`:
```

console.ts:123 [Extension Host] ApplicationInsights:Sender [Object]
t.log @ console.ts:123
t._logExtensionHostMessage @ extensionHost.ts:391
(anonymous) @ extensionHost.ts:211
emitTwo @ events.js:106
emit @ events.js:194
process.nextTick @ internal/child_process.js:766
_combinedTickCallback @ internal/process/next_tick.js:73
_tickCallback @ internal/process/next_tick.js:104
```
Apologies for the confusion I'd completely forgotten that existed.

**Comment by ramya-rao-a** [comment]
Ok, those logs don't help much.. 

I cloned your example from https://github.com/nathj07/go-resourcesync and tried removing the imports (both std lib and the others). They got added back each time on save :)

You mentioned in https://github.com/Microsoft/vscode-go/issues/1427#issuecomment-354743219 that you see the issue only in complex projects. The only reason this may happen is if formatting is taking more than 750ms. 

Can you update to the latest version of the Go extension (0.6.71) and try editing and saving your file again? If the formatter is indeed slow, then the console (Help -> Toggle Developer Tools -> Console) will show a log `Formatting took too long(...ms). Format On Save feature could be aborted.`

**Comment by nathj07** [comment]
On a more complex project I tried with the new version 0.6.71 and I see
```
[Extension Host] Formatting took too long(13734ms). Format On Save feature could be aborted.
```
in that console output.

**Comment by ramya-rao-a** [comment]
Ok, then the issue is not with std lib pkgs vs non std lib pkgs

Its about slow formatters which end up aborting the format on save process which I have covered in https://github.com/Microsoft/vscode-go/issues/1419#issuecomment-355654879

**Comment by nathj07** [comment]
Ok, I guess this can be closed then and the issue dealt with over on #1419
